public/app/features/dimensions/editors/ResourcePickerPopover.tsx

```diff
@@ -0,0 +1,273 @@
+        .then((ds) => {
+          (ds as GrafanaDatasource).listFiles(folder).subscribe({
+            next: (frame) => {
+              const cards: ResourceItem[] = [];
+              frame.forEach((item) => {
+                if (filter(item)) {
+                  const idx = item.name.lastIndexOf('.');
+                  cards.push({
+                    value: `${folder}/${item.name}`,
+                    label: item.name,
+                    search: (idx ? item.name.substr(0, idx) : item.name).toLowerCase(),
+                    imgUrl: `public/${folder}/${item.name}`,
+                  });
+                }
+              });
+              setDirectoryIndex(cards);
+              setFilteredIndex(cards);
+            },
+          });
+        });
+    }
+  }, [mediaType, currentFolder]);
+
+  const [searchQuery, setSearchQuery] = useState<string>();
+  const onChangeSearch = (query: string) => {
+    if (query) {
+      query = query.toLowerCase();
+      setFilteredIndex(directoryIndex.filter((card) => card.search.includes(query)));
+    } else {
+      setFilteredIndex(directoryIndex);
+    }
+  };
+
+  const [newValue, setNewValue] = useState<string>(value ?? '');
+
+  const imgSrc = getPublicOrAbsoluteUrl(newValue!);
+
+  let shortName = newValue?.substring(newValue.lastIndexOf('/') + 1, newValue.lastIndexOf('.'));
+  if (shortName.length > 20) {
+    shortName = shortName.substring(0, 20) + '...';
+  }
+
+  const [activePicker, setActivePicker] = useState<PickerType>('folder');
+  const getTabClassName = (tabName: PickerType) => {
+    return `ResourcePickerPopover__tab ${activePicker === tabName && 'ResourcePickerPopover__tab--active'}`;
+  };
+
+  const theme = useTheme2();
+  const styles = getStyles(theme);
+
+  const renderFolderPicker = () => (
+    <>
+      <Field>
+        <Select menuShouldPortal={true} options={folders} onChange={setCurrentFolder} value={currentFolder} />
+      </Field>
+      <Field>
+        <FilterInput
+          value={searchQuery ?? ''}
+          placeholder="Search"
+          onChange={(v) => {
+            onChangeSearch(v);
+            setSearchQuery(v);
+          }}
+        />
+      </Field>
+      {filteredIndex && (
+        <div className={styles.cardsWrapper}>
+          <ResourceCards cards={filteredIndex} onChange={(v) => setNewValue(v)} value={newValue} />
+        </div>
+      )}
+    </>
+  );
+
+  const renderURLPicker = () => (
+    <>
+      <Field>
+        <Input onChange={(e) => setNewValue(e.currentTarget.value)} value={newValue} />
+      </Field>
+      <div className={styles.iconContainer}>
+        <Field label="Preview">
+          <div className={styles.iconPreview}>
+            {mediaType === 'icon' && <SVG src={imgSrc} className={styles.img} />}
+            {mediaType === 'image' && newValue && <img src={imgSrc} className={styles.img} />}
+          </div>
+        </Field>
+        <Label>{shortName}</Label>
+      </div>
+    </>
+  );
```

Comment:
Maybe these should be a separate components for the sake of this component complexity?